src/stan/lang/ast/var_origin.hpp

```diff
@@ -49,11 +53,91 @@ namespace stan {
                  const bool& is_local);
 
       /**
-       * Return true if declared in a local block.
+       * Return true when declared in data block.
+       *
+       * @return true for data origin block types
+       */
+      bool is_data_origin() const;
+
+      /**
+       * Return true when declared in top-level of parameter or
+       * transformed parameter block.
+       *
+       * @return true for parameter origin block types
+       */
+      bool is_parameter_origin() const;
+
+      /**
+       * Return true when declared in top-level parameter block
+       *
+       * @return true for top-level parameter block
+       */
+      bool is_non_local_parameter_origin() const;
+
+      /**
+       * Return true when declared in top-level transformed parameter block
+       *
+       * @return true for top-level transformed parameter block
+       */
+      bool is_non_local_transformed_parameter_origin() const;
+      /**
+       * Return false when declared in transformed parameter block
+       * or local block.
+       *
+       * @return true for non-parameter block types
+       */
+      bool is_non_parameter_origin() const;
+
+      /**
+       * Return true when declared in any function argument block.
+       *
+       * @return true for function origin block types
+       */
+      bool is_fun_origin() const;
+
+      /**
+       * Return true when declared in void_function_argument_origin block.
+       *
+       * @return true for void function origin block types
+       */
+      bool is_void_function_origin() const;
+
+      /**
+       * Return true when enclosing block is void function type
+       *
+       * @return true for void function origin block types
+       */
+      bool is_non_void_function_origin() const;
+
+      /**
+       * Return true when program block allows assignment to variables
+       * i.e., not data or parameter block
+       *
+       * @return true when program block allows access to LP
+       */
+      bool allows_assignment() const;
+
+      /**
+       * Return true when program block allows access to LP function
+       *
+       * @return true when program block allows access to LP function
+       */
+      bool allows_lp_fun() const;
```

Comment:
If you want to bring this up to date, you could replace `lp` with `target`.  The `lp` will eventually all go away.  I think we still need to make the function suffix `_target` in the language and deprecate `_lp`.  As you pointed out before, it's easily confused with `_lpdf`.